source/adiabatic_conditions/initial_profile.cc

```diff
@@ -187,23 +201,80 @@ namespace aspect
     {
       const double z = this->get_geometry_model().depth(p);
 
+      Assert (z >= 0, ExcInternalError());
+
       if (z >= this->get_geometry_model().maximal_depth())
         {
           Assert (z <= this->get_geometry_model().maximal_depth() + delta_z,
                   ExcInternalError());
           return temperatures.back();
         }
 
-      const unsigned int i = static_cast<unsigned int>(z/delta_z);
-      Assert ((z/delta_z) >= 0, ExcInternalError());
+      if (z < delta_z * std::numeric_limits<double>::epsilon())
+        return temperatures.front();
+
+      const unsigned int i = static_cast<unsigned int>((z/delta_z) * (1. - 2. * std::numeric_limits<double>::epsilon()));
```

Comment:
We are now repeating this logic many times in this file. How about moving this into a helper function?